_pages/typecheck.md

```diff
@@ -684,3 +684,371 @@ Cyclic module dependencies can cause problems for the type checker.  In order to
 ```lua
 local myModule = require(MyModule) :: any
 ```
+
+## Type functions
+
+Type functions are functions that operate on types instead of runtime values. They can use the [types](#types-library) library to transform existing types or create new ones.
+
+Many common type functions are available through the community [TypeForge](https://github.com/cameronpcampbell/typeforge) library.
+
+Here's a simplified implementation of the builtin type function `keyof`. It takes a table type and returns its property names as a [union](typecheck#union-types) of [singletons](typecheck#singleton-types-aka-literal-types).
+
+```luau
+type function simple_keyof(ty)
+    -- Ignoring unions or intersections of tables for simplicity.
+    if not ty:is("table") then
+        error("Can only call keyof on tables.")
+    end
+
+    local union = nil
+
+    for property in ty:properties() do
+        union = if union then types.unionof(union, property) else property
+    end
+
+    return if union then union else types.singleton(nil)
+end
+
+type person = {
+    name: string,
+    age: number,
+}
+--- keys = "age" | "name"
+type keys = simple_keyof<person>
+```
+
+### Type function envrionment
+
+In addition to the [types](#types-library) library, type functions have access to:
+
+* `assert`, `errror`, `print`
+* `next`, `ipairs`, `pairs`
+* `select`, `unpack`
+* `getmetatable`, `setmetatable`
+* `rawget`, `rawset`, `rawlen`, `raweq`
+* `tonumber`, `tostring`
+* `type`, `typeof`
+* `math` library
+* `table` library
+* `string` library
+* `bit32` library
+* `utf8` library
+* `buffer` library
+
+### Type function limitations
+
+Limitations may be lifted in the future.
+
+1. Can't access user-defined types.
+2. Can only access type functions from the same module.
+3. No typechecking for type function implementations.
+4. Can't accept or create generic arguments.
```

Comment:
I'm not sure what you mean here at all.
If you are talking about generic function then this limitation is not correct.